File: VB/CustomWizardExample/CustomWizardCommandHandler.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Diagnostics
Imports System.Windows.Forms
Imports System.Windows.Forms.Design
Imports System.Xml.Linq
Imports CustomWizardExample.Wizard
Imports CustomWizardExample.Wizard.Model
Imports DevExpress.Data.Entity
Imports DevExpress.DataAccess.EntityFramework.Native
Imports DevExpress.DataAccess.Native
Imports DevExpress.DataAccess.Wizard
Imports DevExpress.DataAccess.Wizard.Native
Imports DevExpress.DataAccess.Wizard.Services
Imports DevExpress.Entity.ProjectModel
Imports DevExpress.LookAndFeel
Imports DevExpress.LookAndFeel.DesignService
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraReports.UserDesigner
Imports DevExpress.XtraReports.Wizards3.Builder

Namespace CustomWizardExample
    Public Class CustomWizardCommandHandler
        Implements ICommandHandler

        Private mdiController As XRDesignMdiController

        Public Sub New(ByVal mdiController As XRDesignMdiController)
            Me.mdiController = mdiController
        End Sub

#Region "Implementation of ICommandHandler"

        Public Sub HandleCommand(ByVal command As ReportCommand, ByVal args() As Object) Implements ICommandHandler.HandleCommand
            Select Case command
                Case ReportCommand.NewReportWizard
                    OnNewReportWizard()
                Case ReportCommand.AddNewDataSource
                    OnAddNewDataSource()
            End Select
        End Sub

        Public Function CanHandleCommand(ByVal command As ReportCommand, ByRef useNextHandler As Boolean) As Boolean Implements ICommandHandler.CanHandleCommand
            useNextHandler = Not (command = ReportCommand.NewReportWizard OrElse command = ReportCommand.AddNewDataSource)
            Return Not useNextHandler
        End Function

#End Region

        Private Sub OnNewReportWizard()
            mdiController.OpenReport(New XtraReport())

            Dim designerHost As IDesignerHost = TryCast(mdiController.ActiveDesignPanel.GetService(GetType(IDesignerHost)), IDesignerHost)
            Debug.Assert(designerHost IsNot Nothing, "host != null")

            Dim runner As CustomReportWizardRunner = CreateCustomReportWizardRunner(designerHost)
            Dim client As DataSourceWizardClient = CreateDataSourceWizardClient(designerHost)

            If runner.Run(client) <> DialogResult.OK Then
                Return
            End If

            mdiController.Form.BeginInvoke(New MethodInvoker(Sub()
                                                                 Dim undoEngine As UndoEngine = DirectCast(designerHost.GetService((GetType(UndoEngine))), UndoEngine)
                                                                 Dim myMethod As System.Action = Sub()
                                                                                                     Dim builder As New XtraWizardReportBuilderWrapper(designerHost)
                                                                                                     builder.Build(runner.WizardModel, True, CreateDataComponent(runner.WizardModel))
                                                                                                 End Sub
                                                                 undoEngine.ExecuteWithoutUndo(myMethod)
                                                             End Sub))
        End Sub

        Private Sub OnAddNewDataSource()
            Dim designerHost As IDesignerHost = TryCast(mdiController.ActiveDesignPanel.GetService(GetType(IDesignerHost)), IDesignerHost)
            Debug.Assert(designerHost IsNot Nothing, "host != null")

            Dim runner As CustomDataSourceWizardRunner = CreateCustomDataSourceWizardRunner(designerHost)
            Dim client As DataSourceWizardClient = CreateDataSourceWizardClient(designerHost)

            If runner.Run(client) <> DialogResult.OK Then
                Return
            End If

            Dim dataComponent = CreateDataComponent(runner.WizardModel)

            If TypeOf dataComponent Is IComponent Then
                Dim designerTransaction As DesignerTransaction = designerHost.CreateTransaction()
                designerHost.Container.Add(DirectCast(dataComponent, IComponent))
                designerTransaction.Commit()
            End If
            Dim report As XtraReport = DirectCast(designerHost.RootComponent, XtraReport)
            If report.DataSource Is Nothing Then
                TypeDescriptor.GetProperties(report)("DataSource").SetValue(report, dataComponent)
                TypeDescriptor.GetProperties(report)("DataMember").SetValue(report, dataComponent.DataMember)
            End If
        End Sub

        Private Shared Function CreateDataSourceWizardClient(ByVal host As IDesignerHost) As DataSourceWizardClient
            Dim solutionTypesProvider As ISolutionTypesProvider = EntityServiceHelper.GetRuntimeSolutionProvider(System.Reflection.Assembly.GetEntryAssembly())
            Dim connectionStringsProvider As IConnectionStringsProvider = New RuntimeConnectionStringsProvider()
            Dim connectionService As IConnectionStorageService = New ConnectionStorageService()
            Dim parameterService As IParameterService = DirectCast(host.GetService(GetType(IParameterService)), IParameterService)
            Return New DataSourceWizardClient(connectionService, parameterService, solutionTypesProvider, connectionStringsProvider)
        End Function

        Private Shared Function CreateCustomReportWizardRunner(ByVal host As IDesignerHost) As CustomReportWizardRunner
            Dim lookAndFeel As UserLookAndFeel = DirectCast(host.GetService(GetType(ILookAndFeelService)), ILookAndFeelService).LookAndFeel
            Dim uiService As IUIService = DirectCast(host.GetService(GetType(IUIService)), IUIService)
            Dim owner As IWin32Window = If(uiService IsNot Nothing, uiService.GetDialogOwnerWindow(), Nothing)

            Return New Wizard.CustomReportWizardRunner(lookAndFeel, owner)
        End Function

        Private Shared Function CreateCustomDataSourceWizardRunner(ByVal host As IDesignerHost) As CustomDataSourceWizardRunner
            Dim lookAndFeel As UserLookAndFeel = DirectCast(host.GetService(GetType(ILookAndFeelService)), ILookAndFeelService).LookAndFeel
            Dim uiService As IUIService = DirectCast(host.GetService(GetType(IUIService)), IUIService)
            Dim owner As IWin32Window = If(uiService IsNot Nothing, uiService.GetDialogOwnerWindow(), Nothing)

            Return New Wizard.CustomDataSourceWizardRunner(lookAndFeel, owner)
        End Function

        Private Shared Function CreateDataComponent(ByVal model As CustomReportModel) As IDataComponent
            If model.DataSourceType Is CustomDataSourceType.Persons Then
                If model.PersonsDataSourceType = PersonsDataSourceType.Xml Then
                    Dim root = System.Xml.Linq.XDocument.Load("Data.xml").Root
                    Dim persons As New List(Of Person)()
                    For Each item In root.Elements()
                        Dim firstName = item.GetAttributeValue("FirstName")
                        Dim secondName = item.GetAttributeValue("SecondName")

                        Dim person = New Person() With {.FirstName = firstName, .SecondName = secondName}
                        persons.Add(person)
                    Next item
                    Return New ListDataSource(persons)
                Else
                    Dim staticPersons = New List(Of Person)({
                     New Person() With {.FirstName = "John", .SecondName = "Small"},
                     New Person() With {.FirstName = "Paul", .SecondName = "Green"},
                     New Person() With {.FirstName = "George", .SecondName = "White"}
                    })
                    Return New ListDataSource(staticPersons)
                End If
            End If
            Return model.CreateDataComponent()
        End Function
    End Class
End Namespace

File: CS/CustomWizardExample/CustomWizardCommandHandler.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Diagnostics;
using System.Windows.Forms;
using System.Windows.Forms.Design;
using System.Xml.Linq;
using CustomWizardExample.Wizard;
using CustomWizardExample.Wizard.Model;
using DevExpress.Data.Entity;
using DevExpress.DataAccess.EntityFramework.Native;
using DevExpress.DataAccess.Native;
using DevExpress.DataAccess.Wizard;
using DevExpress.DataAccess.Wizard.Native;
using DevExpress.DataAccess.Wizard.Services;
using DevExpress.Entity.ProjectModel;
using DevExpress.LookAndFeel;
using DevExpress.LookAndFeel.DesignService;
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.UserDesigner;
using DevExpress.XtraReports.Wizards3.Builder;

namespace CustomWizardExample {
    public class CustomWizardCommandHandler : ICommandHandler {
        XRDesignMdiController mdiController;

        public CustomWizardCommandHandler(XRDesignMdiController mdiController) {
            this.mdiController = mdiController;
        }

        #region Implementation of ICommandHandler

        public void HandleCommand(ReportCommand command, object[] args) {
            switch(command) {
                case ReportCommand.NewReportWizard:
                    OnNewReportWizard();
                    break;
                case ReportCommand.AddNewDataSource:
                    OnAddNewDataSource();
                    break;
            }
        }

        public bool CanHandleCommand(ReportCommand command, ref bool useNextHandler) {
            useNextHandler = !(command == ReportCommand.NewReportWizard || command == ReportCommand.AddNewDataSource);
            return !useNextHandler;
        }

        #endregion

        void OnNewReportWizard() {
            mdiController.OpenReport(new XtraReport());

            IDesignerHost designerHost = mdiController.ActiveDesignPanel.GetService(typeof(IDesignerHost)) as IDesignerHost;
            Debug.Assert(designerHost != null, "host != null");

            CustomReportWizardRunner runner = CreateCustomReportWizardRunner(designerHost);
            DataSourceWizardClient client = CreateDataSourceWizardClient(designerHost);

            if(runner.Run(client) != DialogResult.OK)
                return;

            mdiController.Form.BeginInvoke(new MethodInvoker(() => {
                UndoEngine undoEngine = (UndoEngine)designerHost.GetService((typeof(UndoEngine)));
                System.Action myMethod = delegate() { new WizardReportBuilder(designerHost).Build(runner.WizardModel, true, CreateDataComponent(runner.WizardModel)); };
                undoEngine.ExecuteWithoutUndo(myMethod);
            }));
        }
        public void MyMethod(IDesignerHost designerHost, CustomReportWizardRunner runner) {
            new WizardReportBuilder(designerHost).Build(runner.WizardModel, true, CreateDataComponent(runner.WizardModel));
        }
        void OnAddNewDataSource() {
            IDesignerHost designerHost = mdiController.ActiveDesignPanel.GetService(typeof(IDesignerHost)) as IDesignerHost;
            Debug.Assert(designerHost != null, "host != null");

            CustomDataSourceWizardRunner runner = CreateCustomDataSourceWizardRunner(designerHost);
            DataSourceWizardClient client = CreateDataSourceWizardClient(designerHost);

            if(runner.Run(client) != DialogResult.OK)
                return;

            var dataComponent = CreateDataComponent(runner.WizardModel);

            if(dataComponent is IComponent) {
                DesignerTransaction designerTransaction = designerHost.CreateTransaction();
                designerHost.Container.Add((IComponent)dataComponent);
                designerTransaction.Commit();
            }
            XtraReport report = (XtraReport)designerHost.RootComponent;
            if(report.DataSource == null) {
                TypeDescriptor.GetProperties(report)["DataSource"].SetValue(report, dataComponent);
                TypeDescriptor.GetProperties(report)["DataMember"].SetValue(report, dataComponent.DataMember);
            }
        }

        static DataSourceWizardClient CreateDataSourceWizardClient(IDesignerHost host) {
            ISolutionTypesProvider solutionTypesProvider =
                EntityServiceHelper.GetRuntimeSolutionProvider(System.Reflection.Assembly.GetEntryAssembly());
            IConnectionStringsProvider connectionStringsProvider = new RuntimeConnectionStringsProvider();
            IConnectionStorageService connectionService = new ConnectionStorageService();
            IParameterService parameterService = (IParameterService)host.GetService(typeof(IParameterService));
            return new DataSourceWizardClient(connectionService, parameterService, solutionTypesProvider,
                connectionStringsProvider);
        }

        static CustomReportWizardRunner CreateCustomReportWizardRunner(IDesignerHost host) {
            UserLookAndFeel lookAndFeel =
                ((ILookAndFeelService)host.GetService(typeof(ILookAndFeelService))).LookAndFeel;
            IUIService uiService = (IUIService)host.GetService(typeof(IUIService));
            IWin32Window owner = uiService != null ? uiService.GetDialogOwnerWindow() : null;

            return new Wizard.CustomReportWizardRunner(lookAndFeel, owner);
        }

        static CustomDataSourceWizardRunner CreateCustomDataSourceWizardRunner(IDesignerHost host) {
            UserLookAndFeel lookAndFeel =
                ((ILookAndFeelService)host.GetService(typeof(ILookAndFeelService))).LookAndFeel;
            IUIService uiService = (IUIService)host.GetService(typeof(IUIService));
            IWin32Window owner = uiService != null ? uiService.GetDialogOwnerWindow() : null;

            return new Wizard.CustomDataSourceWizardRunner(lookAndFeel, owner);
        }

        static IDataComponent CreateDataComponent(CustomReportModel model) {
            if(model.DataSourceType == CustomDataSourceType.Persons) {
                if(model.PersonsDataSourceType == PersonsDataSourceType.Xml) {
                    var root = System.Xml.Linq.XDocument.Load("Data.xml").Root;
                    List<Person> persons = new List<Person>();
                    foreach(var item in root.Elements()) {
                        var firstName = item.GetAttributeValue("FirstName");
                        var secondName = item.GetAttributeValue("SecondName");

                        var person = new Person() { FirstName = firstName, SecondName = secondName };
                        persons.Add(person);
                    }
                    return new ListDataSource(persons);
                }
                else {
                    var staticPersons = new List<Person>(new[] { 
                        new Person() { FirstName = "John", SecondName = "Small" }, 
                        new Person() { FirstName = "Paul", SecondName = "Green" },
                        new Person() { FirstName = "George", SecondName = "White" }});
                    return new ListDataSource(staticPersons);
                }
            }
            return model.CreateDataComponent();
        }
    }
}
